In [4]:
import requests
from bs4 import BeautifulSoup
url = "https://zh.wikipedia.org/zh-tw/2024%E5%B9%B4%E7%94%B5%E5%BD%B1"
resp = requests.get(url)
resp.encoding = "utf-8"
#print(resp.text)
soup = BeautifulSoup(resp.text,"html5lib")
movieTable = soup.find("table",class_="wikitable sortable")
tbody = movieTable.find("tbody")
trList = tbody.find_all("tr")
print(trList)

[<tr>
<th>排名</th>
<th>片名</th>
<th>發行商</th>
<th>全球票房
</th></tr>, <tr>
<th>1
</th>
<td style="background:#b6fcb6;">《<a href="/wiki/%E8%85%A6%E7%AD%8B%E6%80%A5%E8%BD%89%E5%BD%8E2" title="腦筋急轉彎2">腦筋急轉彎2</a>》†
</td>
<td><a href="/wiki/%E8%8F%AF%E7%89%B9%E8%BF%AA%E5%A3%AB%E5%B0%BC%E5%B7%A5%E4%BD%9C%E5%AE%A4%E9%9B%BB%E5%BD%B1" title="華特迪士尼工作室電影">迪士尼</a>
</td>
<td>$1,687,299,613
</td></tr>, <tr>
<th>2
</th>
<td style="background:#b6fcb6;">《<a href="/wiki/%E6%AD%BB%E4%BE%8D%E8%88%87%E9%87%91%E9%8B%BC%E7%8B%BC" title="死侍與金鋼狼">死侍與金鋼狼</a>》†
</td>
<td><a href="/wiki/%E8%8F%AF%E7%89%B9%E8%BF%AA%E5%A3%AB%E5%B0%BC%E5%B7%A5%E4%BD%9C%E5%AE%A4%E9%9B%BB%E5%BD%B1" title="華特迪士尼工作室電影">迪士尼</a>
</td>
<td>$1,318,568,740
</td></tr>, <tr>
<th>3
</th>
<td style="background:#b6fcb6;">《<a href="/wiki/%E7%A5%9E%E5%81%B7%E5%A5%B6%E7%88%B84" title="神偷奶爸4">神偷奶爸4</a>》†
</td>
<td><a href="/wiki/%E7%8E%AF%E7%90%83%E5%BD%B1%E4%B8%9A" title="環球影業">環球影業</a>
</td>
<td>$948,719,095
</td></tr>, <tr>
<th>4
</th>
<td>《<a href="/wi

In [8]:
import requests
from bs4 import BeautifulSoup
url = "https://zh.wikipedia.org/zh-tw/2024%E5%B9%B4%E7%94%B5%E5%BD%B1"
resp = requests.get(url)
resp.encoding = "utf-8"
#print(resp.text)
soup = BeautifulSoup(resp.text,"html5lib")
movieTable = soup.find("table",class_="wikitable sortable")
tbody = movieTable.find("tbody")
trList = tbody.find_all("tr")
#print(trList)
for i in range(1,len(trList)):
    print("排名:",trList[i].find("th").text,end="")
    print("片名:",trList[i].select_one("a").text)
    print("全球票房:",trList[i].find_all("td")[-1].text)

排名: 1
片名: 腦筋急轉彎2
全球票房: $1,687,299,613

排名: 2
片名: 死侍與金鋼狼
全球票房: $1,318,568,740

排名: 3
片名: 神偷奶爸4
全球票房: $948,719,095

排名: 4
片名: 沙丘：第二部
全球票房: $711,844,358

排名: 5
片名: 哥吉拉與金剛：新帝國
全球票房: $567,650,016

排名: 6
片名: 功夫熊貓4
全球票房: $549,141,851

排名: 7
片名: 熱辣滾燙
全球票房: $479,597,304

排名: 8
片名: 飛馳人生2
全球票房: $466,930,272

排名: 9
片名: 抓娃娃
全球票房: $445,027,050[3][4]

排名: 10
片名: 絕地戰警：生死與共
全球票房: $404,080,232



## Numpy
+ 可做到的功能
  + 快速的多維陣列操作。
  + 可靠的科學功能函數庫。
  + 可協助繪圖工具進行繪圖。
+ NumPy 是幾乎每個科學 Python 應用程序或模組的核心，它提供了以向量化形式的快速 N-d 陣列類型可供操作。

In [12]:
import numpy as np
a = [1,2,3,5,7,9]
b = [3,5,3,6,7,9]
print(type(a))
np1 = np.array(a)
np2 = np.array(b)
print(type(np1))
print(np1)
c = a + b
print(c)
np3 = np1 + np2
print(np3)

<class 'list'>
<class 'numpy.ndarray'>
[1 2 3 5 7 9]
[1, 2, 3, 5, 7, 9, 3, 5, 3, 6, 7, 9]
[ 4  7  6 11 14 18]


+ ndarray.ndim
   + 維度的數量。
+ ndarray.shape
   + 顯示出陣列在每個維度上的整數值。
+ ndarray.size
   + 陣列內元素的總數。
+ ndarray.dtype
   + 用來描述陣列中元素類型的對象。
 
https://www.runoob.com/numpy/numpy-dtype.html

In [21]:
import numpy as np
i = [[1,2,3],
     [4,5,6]]
a = np.array(i,dtype=np.uint16)
print(a.ndim)
#軸 axis
print(a.shape)
print(a.size)
print(a.dtype)

2
(2, 3)
6
uint16


In [30]:
import numpy as np
a = np.array([6,7,8,9])
b = np.arange(4)
print(a)
print(b)
c = a - b
print(c)
d = b ** 2
print(d)
f = np.array([5,-1,3,9,0])
print(f <= 3)
print(f[f <= 3])
f[f <= 3] = 1
print(f)

[6 7 8 9]
[0 1 2 3]
[6 6 6 6]
[0 1 4 9]
[False  True  True False  True]
[-1  3  0]
[5 1 1 9 1]


In [39]:
import numpy as np
x = np.empty((2,4))
print(x)
print("==============================")
x = np.zeros((2,4))
print(x)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]]
[[0. 0. 0. 0.]
 [0. 0. 0. 0.]]


In [46]:
x = np.ones((2,4))
print(x)
x = np.full((2,4),5)
print(x)

[[1. 1. 1. 1.]
 [1. 1. 1. 1.]]
[[5 5 5 5]
 [5 5 5 5]]


In [64]:
import numpy as np
x = np.eye(3)
print(x)
x = np.diag([1,2,3])
print("x:",x)

[[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
x: [[1 0 0]
 [0 2 0]
 [0 0 3]]


## 陣列初始化與重新規劃
+ 針對初始化這部分提供了多種方式：
+ 不做初始化
   + 例如：np.empty((2,4))
+ 全部規劃為 0
   + 例如：np.zeros((3,4))
+　全部規劃為 1
  + np.ones((5,2))
+ 全部規劃／填滿為某數，例如 6
   + 例如： np.full((3,3),6)
+ 建立對角矩陣，對角皆為 1，其他為 0
   + 例如： np.eye(3)
+ 建立對角矩陣，對角為 1,2,3,4，其他為 0
   + 例如： np.diag([1,2,3,4])
+ reshape((x,y)) 這個指令就可以將現有的陣列  重新規劃為 x 乘以 y 的陣列。 
+ np.arange(起始值, 結束值, 固定間隔)：也是產生一維陣列，和 np.array( ) 的差別在於 arange 擁有較大的彈性，而且元素數值是自動化產生。
+ np.linspace(起始值, 結束值, 起始與結束的區間內要產生幾個元素 )：只要給定陣列的區間(起始值、結束值)，就可以要求在這個區間內產生幾個元素。
+ arange(a,b,c)
   + 不包含 b 由 a 開始 以固定間隔 c 來取值
+ linspace(a,b,c,)
   + 包含 b 由 a 開始 切割成 c 個資料
+ linspace(a,b,c,endpoint=False)
   + 不包含 b 由 a 開始 切割成 c 個資料

### linspace arange

In [86]:
import numpy as np
x = np.linspace(1,10,6)
print(x)
y = np.linspace(1,10,6,endpoint=False)
print(y)

x = np.linspace(2,20,10).reshape((5,2))
print(x)

x = np.arange(8).reshape((4,2))
print(x)


[ 1.   2.8  4.6  6.4  8.2 10. ]
[1.  2.5 4.  5.5 7.  8.5]
[[ 2.  4.]
 [ 6.  8.]
 [10. 12.]
 [14. 16.]
 [18. 20.]]
[[0 1]
 [2 3]
 [4 5]
 [6 7]]


In [89]:
import numpy as np
r1 = np.arange(25,30,.5)
print(r1)

[25.  25.5 26.  26.5 27.  27.5 28.  28.5 29.  29.5]


In [96]:
# 3~5 取9切分段落
import numpy as np
lin = np.linspace(3,5,9)
print(lin)
print(lin.shape)

[3.   3.25 3.5  3.75 4.   4.25 4.5  4.75 5.  ]
(9,)


## 陣列指定位置取值
+ 間隔選取 [::c]
  + 以 1 維陣列來說明 x[a:b:c]
  + a：選取資料的起始索引
  + b：選取資料的結束索引 +1
  + c：選取資料間隔，以索引值可以被此值整除的元素，不指定表示 1
+ 倒序 [::-1]
  + 只是單純的把順序反過來

## 陣列指定位置-給予一個整數
+ 關於指定位置  [row,column]
+ 假設給予一個整數為 N
 + 如果是給固定的 N，那就代表 row 或 column 等於 N。
 + 如果是 N：，那就代表 row 或 column 大於等於 N 的區域。
 + 如果是：N，那就代表 row 或 column 小於 N 的區域。 
 + 如果是：，那就代表 row 或 column 是任意欄位。

In [102]:
import numpy as np
a = np.arange(1,10).reshape((3,3))
print(a)
print(a[0])
print(a[2][1])
print(a[2,1])
print(a[1,2])

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[1 2 3]
8
8
6


In [110]:
import numpy as np
a = np.arange(1,10).reshape((3,3))
print(a)
print("="*30)
print(a[1,1:3])
print("="*30)
print(a[:,1])
print("="*30)
a[1,2] = 75
print(a)
print("="*30)
a[:,0] = [100,101,102]
print("="*30)
print(a)
print("="*30)
a[1,:] = [200,201,202] 
print(a)

[[1 2 3]
 [4 5 6]
 [7 8 9]]
[5 6]
[2 5 8]
[[ 1  2  3]
 [ 4  5 75]
 [ 7  8  9]]
[[100   2   3]
 [101   5  75]
 [102   8   9]]
[[100   2   3]
 [200 201 202]
 [102   8   9]]


In [115]:
import numpy as np
a = np.arange(36).reshape((6,6))
print(a)
print("====1====")
print(a[0,3:5])
print("====2====")
print(a[4:,4:])
print("====3===")
print(a[:3,:3])
print("====4===")
print(a[:,2])

[[ 0  1  2  3  4  5]
 [ 6  7  8  9 10 11]
 [12 13 14 15 16 17]
 [18 19 20 21 22 23]
 [24 25 26 27 28 29]
 [30 31 32 33 34 35]]
====1====
[3 4]
====2====
[[28 29]
 [34 35]]
====3===
[[ 0  1  2]
 [ 6  7  8]
 [12 13 14]]
====4===
[ 2  8 14 20 26 32]


## 重設新陣列
+ resize 這個動作會依據原本的陣列再設定指定大小的新陣列。 
+ resize 動作的參數如下：
+ numpy.resize(arr1，shape1)
  + arr1：原本的陣列
  + shape1：新規劃的大小
+ 如果規劃的新陣列比較大，將會重新複製原有陣列的資料，填滿新的儲存格。
+ resize 這個動作***建立新的陣列***，而 reshape 則是依據原有的重新規劃，仍會受到原有陣列影響。
+ transpose()轉置

In [123]:
import numpy as np
a = np.array([[1,2,3],
              [4,5,6]])
print(a)
print(a.shape)
b = a.reshape((1,6))
print(b)
print(b.shape)
b[0,1] = 100
print(a)# 會改變
print(b)

[[1 2 3]
 [4 5 6]]
(2, 3)
[[1 2 3 4 5 6]]
(1, 6)
[[  1 100   3]
 [  4   5   6]]
[[  1 100   3   4   5   6]]


### resize

In [128]:
import numpy as np
a = np.array([[1,2,3],
              [4,5,6]])
print(a)
print(a.shape)
b = np.resize(a,(3,2))
print(b)
print(b.shape)
b[1,1] = 100
print(b)
print(a)#不會變

[[1 2 3]
 [4 5 6]]
(2, 3)
[[1 2]
 [3 4]
 [5 6]]
(3, 2)
[[  1   2]
 [  3 100]
 [  5   6]]
[[1 2 3]
 [4 5 6]]


In [133]:
print(a)
c = np.resize(a,(2,4))
print(c)
c = np.resize(a,(1,2))
print(c)
c = np.resize(a,(10,10))
print(c)

[[1 2 3]
 [4 5 6]]
[[1 2 3 4]
 [5 6 1 2]]
[[1 2]]
[[1 2 3 4 5 6 1 2 3 4]
 [5 6 1 2 3 4 5 6 1 2]
 [3 4 5 6 1 2 3 4 5 6]
 [1 2 3 4 5 6 1 2 3 4]
 [5 6 1 2 3 4 5 6 1 2]
 [3 4 5 6 1 2 3 4 5 6]
 [1 2 3 4 5 6 1 2 3 4]
 [5 6 1 2 3 4 5 6 1 2]
 [3 4 5 6 1 2 3 4 5 6]
 [1 2 3 4 5 6 1 2 3 4]]


### transpose

In [139]:
import numpy as np
newArray = np.arange(10).reshape((2,5))
print(newArray)
print("======================")
print(newArray.transpose())
print("======================")
print(newArray)
print("======================")
print(newArray.T)

[[0 1 2 3 4]
 [5 6 7 8 9]]
[[0 5]
 [1 6]
 [2 7]
 [3 8]
 [4 9]]
[[0 1 2 3 4]
 [5 6 7 8 9]]
[[0 5]
 [1 6]
 [2 7]
 [3 8]
 [4 9]]


## 總和最大與最小
+ np.sum( ) 代表某一個陣列內容的總和，也可以指定這個陣列的哪一軸 (axis) 內容總和。
+ np.min( ) 代表某一個陣列內容的最小值，也可以指定這個陣列的哪一軸 (axis) 內容的最小值。
+ np.max( ) 代表某一個陣列內容的最大值，也可以指定這個陣列的哪一軸 (axis) 內容的最大值。
+ amin 與 min 是相同功能的方法。
+ amax 與 max 是相同功能的方法。

### 總和sum()

In [145]:
import numpy as np
a = np.array([[0, 30,45],
              [60,75,90]])
print(a)
print("======================")
b = np.sum(a)
print(b)
print("======================")
b = np.sum(a,axis=0)
print(b)
print("======================")
b = np.sum(a,axis=1)
print(b)
print("======================")

[[ 0 30 45]
 [60 75 90]]
300
[ 60 105 135]
[ 75 225]


### max() min()

In [150]:
import numpy as np
a = np.array([[0, 75,45],
              [60,30,90]])
print(a)
b = np.max(a)
print(b)
print("========")
b = np.max(a,axis = 0)
bi = np.argmax(a,axis = 0)
print(b)
print(bi)
print("========")
b = np.max(a,axis = 1)
bi = np.argmax(a,axis=1)
print(b)
print(bi)


[[ 0 75 45]
 [60 30 90]]
90
[60 75 90]
[1 0 1]
[75 90]
[1 2]


## 中位數與平均
+ median( ) 方法可取出陣列或陣列中指定的軸之中位數。
+ 把所有值高低排序後找出正中間的一個作爲中位數。如果資料數量為偶數，則中位數取最中間的兩個數值的平均數。
+ 參數：
  + array：陣列
  + axis：軸
+ mean( ) 方法可取出陣列或陣列中指定的軸之平均數。
+ 平均值就是把陣列或指定軸的所有資料除以數量後的數值。
+ 參數：
    + array：陣列 
    + axis：軸
+ average( ) 方法與 mean( ) 方法相似，但可以加入權重進行計算。
+ 計算公式為 ((資料*權重)相加)/(權重相加)
+ 參數：
  + array：陣列
  + axis：軸
  + weights：權重，若沒有則設定為 1
  + returned：若設定為 true，代表返回計算結果跟權重相加總合兩筆資料，若沒有設定或設定  false 則只有計算結果 
+ 沒有指定權重時與一般平均值相同。
+ 加權平均值是由每個資料乘以權重以反映加上重要性因素後產生的平均值。
+ average( ) 使用時若沒有指定軸，則陣列將被平坦化。
+ 假設有個陣列資料為 [1,2,3,4]，相應的權重 [40,30,20,10]，加權平均數是這樣：
    (1x40+2x30+3x20+4x10)/(40+30+20+10)

In [153]:
import numpy as np
a = np.array([[0, 30,45],
              [60,75,90]])
print(a)
print( )
b = np.mean(a)
print(b)
print()
b = np.mean(a,axis = 0)
print(b)
print()
b = np.mean(a,axis = 1)
print(b)


[[ 0 30 45]
 [60 75 90]]

50.0

[30.  52.5 67.5]

[25. 75.]


In [161]:
import numpy as np
a = np.array([1,2,3,4])
b = np.average(a)
print(b)
wts = np.array([4,3,2,1])
b = np.average(a,weights=wts)
#(1*4+2*3+3*2+4*1)/(4+3+2+1)  
# 20/10  = 2
print(b)
b ,w = np.average(a,weights=wts,returned=True)
print(b,w)

2.5
2.0
2.0 10.0


## 中位數

In [165]:
import numpy as np
a = np.array([[0, 30,45],
              [60,75,90]])

print(a)
b = np.median(a)
print(b)
print()
b = np.median(a,axis = 0)
print(b)
b = np.median(a,axis = 1)
print(b)

[[ 0 30 45]
 [60 75 90]]
52.5

[30.  52.5 67.5]
[30. 75.]


## 變異數與標準差
+ 變異數代表所有資料到平均數的距離之平方。
+ 標準差代表變異數開平方根，也就是可依此表示資料的分散程度。
+ var( ) 代表由陣列取得變異數的方法。
+ std( ) 代表由陣列取得標準差的方法。

+ 定義：

+ 變異數代表((每一個資料-平均值)平方後的加總結果)/總個數。

+ 標準差是變異數的開根號結果。



In [168]:
import math
import numpy as np
a = np.array([1,2,3,4])
print(a)
b = np.var(a)
print(b)
s = np.std(a)
print(s)
print(math.sqrt(b))

[1 2 3 4]
1.25
1.118033988749895
1.118033988749895


## 平均中位標準差!
#### 請查看男女生遲到天數，要如何進行分析?

In [174]:
import numpy as np
girl=np.array([8,10,15,20,27,29,29,30,30,162])
boy=np.array([34,34,35,35,36,36,36,37,38,39])
g_mean = np.mean(girl)
b_mean = np.mean(boy)
print("g:",g_mean)
print("b:",b_mean)
g_std = np.std(girl)
b_std = np.std(boy)
print(g_std)
print(b_std)
print("======================")
g_median = np.median(girl)
b_median = np.median(boy)
print(g_median)
print(b_median)

g: 36.0
b: 36.0
42.75979419969184
1.5491933384829668
28.0
36.0


## 範圍與百分位數
+ np.ptp( ) 計算最大與最小值的差（最大 -(減) 最小）。
+ np.percentile( ) 方法代表百分位數，用於統計數據的度量指標，需要三個參數：
  + array：陣列。
  + percent：計算的百分位數，介於 0 到 100 之間。
  + axis：進行計算的軸。
+ 百分位數是一種位置量數，有助於瞭解資料在最小值與最大值之間的分布情況。
+ p- 百分位數表示：
  + 至少有 p-百分比 (p%) 的觀察值小於或者等於他。
  + 至少有 (100-p)% 的觀察值大於或等於他。
+ 計算方式：
  + 將資料由小到大排序。
  + 計算 i=(p%*樣本數) 之位置。
  + 於 i 位置：
    + a. 若不是整數，無條件進位取比 i 大的下一個整數位置的值。
    + b. 若為整數，則由 i 與 i+1 兩個位置上的值取平均。

In [178]:
import numpy as np
a = np.array([[0, 30,45],
              [60,75,90]])
print(a)
print()
print("--------------------------")
print(np.ptp(a))
print("--------------------------")
print(np.ptp(a,axis=0))
print("--------------------------")
print(np.ptp(a,axis=1))

[[ 0 30 45]
 [60 75 90]]

--------------------------
90
--------------------------
[60 45 45]
--------------------------
[45 30]


計算百分位數的過程如下：

1. **排序數據**：首先將所有數據排序。
2. **計算位置**：使用公式 `位置 = (n - 1) * (百分位數 / 100)`，其中 `n` 是數據的總數。
3. **內插計算**：如果位置是整數，則百分位數就是該位置的值。如果位置不是整數，則需要使用內插法計算確切的百分位數。


### 原始數據
```
a = [[11, 8, 9], [5, 4, 1]]
```

### 步驟 1：排序數據
將數據展平並排序：
```
[1, 4, 5, 8, 9, 11]
```

### 步驟 2：計算位置
對於第 25 百分位數，計算位置：
```
位置 = (n - 1) * (25 / 100)
    = (6 - 1) * 0.25
    = 5 * 0.25
    = 1.25
```

### 步驟 3：內插計算
位置 1.25 不是整數，所以我們需要進行內插計算。位置 1.25 的值位於位置 1 和位置 2 之間。

位置 1 的值是 4，位置 2 的值是 5。

內插計算公式：
```
百分位數 = 值1 + (值2 - 值1) * 小數部分
         = 4 + (5 - 4) * 0.25
         = 4 + 1 * 0.25
         = 4 + 0.25
         = 4.25
```

因此，第 25 百分位數的值是 4.25。

### 總結
這就是為什麼 `np.percentile(a, 25)` 會輸出 4.25 的原因。計算過程包括排序數據、計算位置，並使用內插法計算確切的百分位數。

In [179]:
import numpy as np
a = np.array([[11, 8, 9],
               [5, 4, 1]])
b1 = np.percentile(a,25)
print(b1)

4.25


計算百分位數的過程如下：

1. **排序數據**：首先將所有數據排序。
2. **計算位置**：使用公式 `位置 = (n - 1) * (百分位數 / 100)`，其中 `n` 是數據的總數。
3. **內插計算**：如果位置是整數，則百分位數就是該位置的值。如果位置不是整數，則需要使用內插法計算確切的百分位數。


### 原始數據
```
a = [[11, 8, 9], [5, 4, 1]]
```

### 沿軸 0 計算第 75 百分位數
沿著軸 0 計算，即對每列（每個 column）分別計算第 75 百分位數。

#### 第 1 列: [11, 5]
1. 排序數據: [5, 11]
2. 計算位置: `(2 - 1) * 0.75 = 1 * 0.75 = 0.75`
3. 內插計算: `5 + (11 - 5) * 0.75 = 5 + 6 * 0.75 = 5 + 4.5 = 9.5`

#### 第 2 列: [8, 4]
1. 排序數據: [4, 8]
2. 計算位置: `(2 - 1) * 0.75 = 1 * 0.75 = 0.75`
3. 內插計算: `4 + (8 - 4) * 0.75 = 4 + 4 * 0.75 = 4 + 3 = 7`

#### 第 3 列: [9, 1]
1. 排序數據: [1, 9]
2. 計算位置: `(2 - 1) * 0.75 = 1 * 0.75 = 0.75`
3. 內插計算: `1 + (9 - 1) * 0.75 = 1 + 8 * 0.75 = 1 + 6 = 7`

### 總結
將各列的第 75 百分位數組合在一起，得到結果：
```
第 75 百分位數 (軸 0): [9.5, 7, 7]
```

這些就是 `np.percentile(a, 75, axis=0)` 的具體算式和結果。

In [180]:
b2= np.percentile(a, 75, axis=0)
print("The percentile is:",b2) 

The percentile is: [9.5 7.  7. ]


### 原始數據
```
a = [[11, 8, 9], [5, 4, 1]]
```

### 沿軸 1 計算第 25 百分位數
我們來逐步進行計算：

#### 第 1 行: [11, 8, 9]
1. 排序數據: [8, 9, 11]
2. 計算位置: `(3 - 1) * 0.25 = 2 * 0.25 = 0.5`
3. 內插計算: 
   - 在 NumPy 中，當位置是一個非整數時，它會在兩個最近的數字之間進行內插。
   - `8 + (9 - 8) * 0.5 = 8 + 1 * 0.5 = 8.5`

#### 第 2 行: [5, 4, 1]
1. 排序數據: [1, 4, 5]
2. 計算位置: `(3 - 1) * 0.25 = 2 * 0.25 = 0.5`
3. 內插計算: 
   - `1 + (4 - 1) * 0.5 = 1 + 3 * 0.5 = 1 + 1.5 = 2.5`

### 總結
將各行的第 25 百分位數組合在一起，得到結果：
```
第 25 百分位數 (軸 1): [8.5, 2.5]
``` 2.5]`，

In [181]:
b3 = np.percentile(a, 25, axis=1)
b3

array([8.5, 2.5])

## 累加與累差
+ np.cumsum( ) 方法進行指定軸資料的累加。
+ 參數：
  + array：陣列
  + axis：軸，沒有指定軸則會平坦化後進行累加
+ 維持原有的資料數量，且在計算結果上進行下一個計算。
+ np.diff( ) 方法進行指定軸資料的累差。
+ 參數：
  + array：陣列
  + axis：軸，沒有指定軸預設 axis 為 1
+ 以原資料進行資料內容刪除，若 axis 為 0 則減少一個 row，若 axis 為 1 減少一個 column。

In [186]:
import numpy as np
a = np.array([[1, 2, 3], 
              [13, 6, 9], 
              [12, 24, 36]])
print(a)
print(np.cumsum(a))
print("======================")
print(np.cumsum(a,axis=0))
print("======================")
print(np.cumsum(a,axis=1))

[[ 1  2  3]
 [13  6  9]
 [12 24 36]]
[  1   3   6  19  25  34  46  70 106]
[[ 1  2  3]
 [14  8 12]
 [26 32 48]]
[[ 1  3  6]
 [13 19 28]
 [12 36 72]]


In [189]:
import numpy as np
a = np.array([[1, 2, 3], 
              [13, 6, 9]])
print(a)
print(np.diff(a))#預設axis=1
print("=================")
print(np.diff(a,axis=0))
print("=================")
print(np.diff(a,axis=1))

[[ 1  2  3]
 [13  6  9]]
[[ 1  1]
 [-7  3]]
[[12  4  6]]
[[ 1  1]
 [-7  3]]
